### PR TITLE
ADD ability to specify policy config yaml as URL

### PR DESCRIPTION
**Pull request by robnester-rh**
This commit adds the ability to specify an https URL for a policy config YAML file.

Resolves: EC-40
Resolves: #964

### CODE REVIEW

File: internal/validate/helpers.go [hunk 1]
```diff
@@ -55,6 +55,32 @@ func GetPolicyConfig(ctx context.Context, policyConfiguration string) (string, e
 		policyConfiguration = configFile
 	}
 
+	// Check if policyConfig is a url. If so, try to download it. If successful we write
+	// that into that into the data.policyConfiguration var.
+	if source.SourceIsHttps(policyConfiguration) {
```
Comment (zregvart) [inline internal/validate/helpers.go:60]: Fair bit of duplication with the code above, perhaps setting a variable to the download function based on those conditionals could help reduce the duplication, something like:

```go
var downloadFn func(context.Context, string, string)
if source.SourceIsGit(policyConfiguration) {
	downloadFn = source.GitConfigDownload
} else if source.SourceIsHttps(policyConfiguration) {
	downloadFn = source.HttpsConfigDownload
}

if downloadFn == nil {
	return "", nil // or something?
}

fs := utils.FS(ctx)
tmpDir, err := utils.CreateWorkDir(fs)
if err != nil {
	return "", err
}
defer utils.CleanupWorkDir(fs, tmpDir)

// Https download and find a suitable config file
configFile, err := downloadFn(ctx, tmpDir, policyConfiguration)
if err != nil {
	return "", err
}
// ...
```

File: internal/policy/source/helpers.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+var policyFileExtensions = []string{
+	"json",
+	"yaml",
+	"yml",
+}
+
+// choosePolicyFile picks a file from a given directory to use as an EC policy file
+// Uses policyFileBaseNames and policyFileExtensions to decide what to look for
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/policy/source/helpers.go:41]: ack

File: internal/validate/helpers.go [hunk 3]
```diff
@@ -42,35 +43,41 @@ func GetPolicyConfig(ctx context.Context, policyConfiguration string) (string, e
 		if err != nil {
 			return "", err
 		}
-		defer utils.CleanupWorkDir(fs, tmpDir)
+		// defer utils.CleanupWorkDir(fs, tmpDir)
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/validate/helpers.go:46]: Nope, thanks for catching!

File: internal/validate/helpers.go [hunk 4]
```diff
@@ -42,35 +43,41 @@ func GetPolicyConfig(ctx context.Context, policyConfiguration string) (string, e
...
 
 		// Git download and find a suitable config file
 		configFile, err := source.GitConfigDownload(ctx, tmpDir, policyConfiguration)
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/validate/helpers.go:49]: I don't disagree. Perhaps it should be made generic to `goGetterDownload`?

File: internal/policy/source/git_config_test.go [hunk 5]
```diff
@@ -19,70 +19,34 @@
...
-				assert.Equal(t, "", ff)
-				assert.ErrorContains(t, err, tt.errText)
-			} else {
-				assert.Equal(t, tt.want, ff)
-				assert.NoError(t, err)
-			}
-		})
+		assert.Equal(t, tt.want, SourceIsFile(tt.src))
+		if tt.want != SourceIsFile(tt.src) {
+			t.Errorf("SourceIsFile(%s) = %v, want %v", tt.src, SourceIsFile(tt.src), tt.want)
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/policy/source/git_config_test.go:48]: Ack. Thanks!

File: internal/policy/source/helpers.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
+// Copyright The Enterprise Contract Contributors
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/policy/source/helpers.go:1]: ack.

File: internal/policy/source/git_config.go [hunk 7]
```diff
@@ -24,22 +24,34 @@ package source
...
 	log "github.com/sirupsen/logrus"
-	"github.com/spf13/afero"
-
-	"github.com/enterprise-contract/ec-cli/internal/utils"
 )
 
+// SourceIsFile returns true if go-getter thinks the src looks like a file path
+func SourceIsFile(src string) bool {
+	normalizedUrl, err := getter.Detect(src, ".", []getter.Detector{new(getter.FileDetector)})
+	return err == nil && strings.HasPrefix(normalizedUrl, "file") && !SourceIsGit(src)
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/policy/source/git_config.go:38]: done

File: internal/validate/helpers.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/validate/helpers.go]: Renamed

File: internal/validate/helpers.go [hunk 8]
```diff
@@ -42,35 +43,41 @@ func GetPolicyConfig(ctx context.Context, policyConfiguration string) (string, e
...
 		if err != nil {
 			return "", err
 		}
-
-		// Changing data.policyConfiguration to the name of the newly downloaded file means we can
-		// use the code below to load the config
-		policyConfiguration = configFile
+		return readPolicyConfigurationFile(ctx, configFile)
+	} else if source.SourceIsFile(policyConfiguration) && utils.HasJsonOrYamlExt(policyConfiguration) {
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/validate/helpers.go:54]: Could you clarify what you mean?

File: internal/policy/source/git_config_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/policy/source/git_config_test.go]: Updated

**Review by lcarva — APPROVED** [review]

File: internal/validate/helpers.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (lcarva) [reply internal/validate/helpers.go:54]: Minor style improvement. All good.

File: internal/policy/source/helpers.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lcarva) [reply internal/policy/source/helpers.go]: Yeah... which is odd :sweat_smile: I would expect a reference to point to an exact file within the repo. Well that's for another day.

**Pull request merged**